In [1]:
import pandas as pd
import numpy as np
import glob

In [2]:
# Read files in data folder to a dataframe
path ='data' # use your path
allFiles = glob.glob(path + "/*.csv")
loans_df = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,skiprows=1,low_memory=False)
    list_.append(df)
loans_df = pd.concat(list_)

In [3]:
loans_df.head(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501        NaN     5000.0       5000.0           4975.0   36 months   
1  1077430        NaN     2500.0       2500.0           2500.0   60 months   
2  1077175        NaN     2400.0       2400.0           2400.0   36 months   
3  1076863        NaN    10000.0      10000.0          10000.0   36 months   
4  1075358        NaN     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade       ...        \
0   10.65%       162.87     B        B2       ...         
1   15.27%        59.83     C        C4       ...         
2   15.96%        84.33     C        C5       ...         
3   13.49%       339.31     C        C1       ...         
4   12.69%        67.79     B        B5       ...         

  hardship_payoff_balance_amount hardship_last_payment_amount  \
0                            NaN                          NaN   
1                            NaN                          NaN   
2                            NaN                          NaN   
3                            NaN                          NaN   
4                            NaN                          NaN   

  disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
0                Cash                     N                       NaN   
1                Cash                     N                       NaN   
2                Cash                     N                       NaN   
3                Cash                     N                       NaN   
4                Cash                     N                       NaN   

  settlement_status settlement_date settlement_amount settlement_percentage  \
0               NaN             NaN               NaN                   NaN   
1               NaN             NaN               NaN                   NaN   
2               NaN             NaN               NaN                   NaN   
3               NaN             NaN               NaN                   NaN   
4               NaN             NaN               NaN                   NaN   

  settlement_term  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 151 columns]

In [4]:
loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321864 entries, 0 to 103547
Columns: 151 entries, id to settlement_term
dtypes: float64(112), object(39)
memory usage: 1.5+ GB


There are total of 1321864 rows and 151 columns in the dataset

In [6]:
#Now lets see how many null values are present in each column
loans_df.isnull().sum().sort_values(ascending=False)

sec_app_open_act_il                           1321864
sec_app_earliest_cr_line                      1321864
sec_app_mort_acc                              1321864
sec_app_open_acc                              1321864
sec_app_revol_util                            1321864
sec_app_fico_range_high                       1321864
sec_app_num_rev_accts                         1321864
sec_app_chargeoff_within_12_mths              1321864
sec_app_collections_12_mths_ex_med            1321864
sec_app_mths_since_last_major_derog           1321864
sec_app_fico_range_low                        1321864
sec_app_inq_last_6mths                        1321864
member_id                                     1321864
revol_bal_joint                               1321864
orig_projected_additional_accrued_interest    1317524
hardship_reason                               1316533
hardship_last_payment_amount                  1316533
hardship_payoff_balance_amount                1316533
hardship_length             

In [5]:
#From the above you can see there are many missing values in each colum/attribute.
#Now we can get rid of the columns which have more than 30% missing or null values
threshold_count = len(loans_df)*0.7 
# Drop any column with more than 30% missing values
filtered_loans_df = loans_df.dropna(thresh=threshold_count,axis=1)

In [8]:
filtered_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321864 entries, 0 to 103547
Data columns (total 93 columns):
id                            1321864 non-null object
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
term                          1321847 non-null object
int_rate                      1321847 non-null object
installment                   1321847 non-null float64
grade                         1321847 non-null object
sub_grade                     1321847 non-null object
emp_title                     1241887 non-null object
emp_length                    1321847 non-null object
home_ownership                1321847 non-null object
annual_inc                    1321843 non-null float64
verification_status           1321847 non-null object
issue_d                       1321847 non-null object
loan_status                   1321847 non-null object
pymnt_plan         

After removing the columns that have more than 30% missing columns, we have total number rows of 1321864. but the columns have reduced to 93.

In [6]:
#Now lets see how many null values are present in each column
filtered_loans_df.isnull().sum().sort_values(ascending=False)

mths_since_recent_inq         182194
num_tl_120dpd_2m              119016
mo_sin_old_il_acct            107552
emp_title                      79977
pct_tl_nvr_dlq                 70446
avg_cur_bal                    70305
mo_sin_old_rev_tl_op           70294
mo_sin_rcnt_rev_tl_op          70294
num_rev_accts                  70294
num_rev_tl_bal_gt_0            70293
num_accts_ever_120_pd          70293
num_actv_bc_tl                 70293
num_actv_rev_tl                70293
num_bc_tl                      70293
num_il_tl                      70293
num_op_rev_tl                  70293
mo_sin_rcnt_tl                 70293
num_tl_30dpd                   70293
num_tl_90g_dpd_24m             70293
num_tl_op_past_12m             70293
tot_hi_cred_lim                70293
total_rev_hi_lim               70293
tot_cur_bal                    70293
total_il_high_credit_limit     70293
tot_coll_amt                   70293
bc_util                        63337
percent_bc_gt_75               63041
b

In [7]:
filtered_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321864 entries, 0 to 103547
Data columns (total 93 columns):
id                            1321864 non-null object
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
term                          1321847 non-null object
int_rate                      1321847 non-null object
installment                   1321847 non-null float64
grade                         1321847 non-null object
sub_grade                     1321847 non-null object
emp_title                     1241887 non-null object
emp_length                    1321847 non-null object
home_ownership                1321847 non-null object
annual_inc                    1321843 non-null float64
verification_status           1321847 non-null object
issue_d                       1321847 non-null object
loan_status                   1321847 non-null object
pymnt_plan         

After filtering the dataset we have 93 columns. Among 93 columns, 67 are of numerical data type and 26 are categorical data types

In [8]:
# Now we will drop some columns which doesn't really affect our analysis
#The cloulms are url,id 
filtered_loans_df = filtered_loans_df.drop(['id' ,'url'],axis=1)

In [9]:
#Drop the rows which has null values in each column
filtered_loans_df = filtered_loans_df.dropna(axis=0,how='all')

In [10]:
filtered_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 91 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
term                          1321847 non-null object
int_rate                      1321847 non-null object
installment                   1321847 non-null float64
grade                         1321847 non-null object
sub_grade                     1321847 non-null object
emp_title                     1241887 non-null object
emp_length                    1321847 non-null object
home_ownership                1321847 non-null object
annual_inc                    1321843 non-null float64
verification_status           1321847 non-null object
issue_d                       1321847 non-null object
loan_status                   1321847 non-null object
pymnt_plan                    1321847 non-null object
purpose            

Now the column count has been reduced to 91 and row count also reduced to 1321847

In [11]:
# Now we will remove the columns that have only one value
filtered_loans_df = filtered_loans_df.loc[:,filtered_loans_df.apply(pd.Series.nunique) != 1]

In [12]:
filtered_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 90 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
term                          1321847 non-null object
int_rate                      1321847 non-null object
installment                   1321847 non-null float64
grade                         1321847 non-null object
sub_grade                     1321847 non-null object
emp_title                     1241887 non-null object
emp_length                    1321847 non-null object
home_ownership                1321847 non-null object
annual_inc                    1321843 non-null float64
verification_status           1321847 non-null object
issue_d                       1321847 non-null object
loan_status                   1321847 non-null object
pymnt_plan                    1321847 non-null object
purpose            

Now the number of columns are reduced to 90

In [13]:
#Now filter the columns that have less than 4 unique values
for col in filtered_loans_df.columns:
    if (len(filtered_loans_df[col].unique()) < 4):
        print(filtered_loans_df[col].value_counts())
        print()

 36 months    944664
 60 months    377183
Name: term, dtype: int64

Source Verified    503870
Verified           417390
Not Verified       400587
Name: verification_status, dtype: int64

n    1321249
y        598
Name: pymnt_plan, dtype: int64

w    764274
f    557573
Name: initial_list_status, dtype: int64

Individual    1312547
Joint App        9300
Name: application_type, dtype: int64

N    1321049
Y        798
Name: hardship_flag, dtype: int64

Cash         1318953
DirectPay       2894
Name: disbursement_method, dtype: int64

N    1307295
Y      14552
Name: debt_settlement_flag, dtype: int64



 There are total 7 columns that have only 2 unique values.
 <p> "term" is an important factor /attribute in any loan. "debt_settlement_flag" can be important to our analysis.so we don't remove term and debt_settlement_flag from the dataset</p>
 <p>But we can drop the rest of the columns </p>


In [14]:
filtered_loans_df = filtered_loans_df.drop(['pymnt_plan','initial_list_status','hardship_flag','disbursement_method' ,'application_type'],axis=1)

In [15]:
filtered_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 85 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
term                          1321847 non-null object
int_rate                      1321847 non-null object
installment                   1321847 non-null float64
grade                         1321847 non-null object
sub_grade                     1321847 non-null object
emp_title                     1241887 non-null object
emp_length                    1321847 non-null object
home_ownership                1321847 non-null object
annual_inc                    1321843 non-null float64
verification_status           1321847 non-null object
issue_d                       1321847 non-null object
loan_status                   1321847 non-null object
purpose                       1321847 non-null object
title              

In [16]:
# Now we will divide our dataset into 2 dataframes. 
#1. Numerical 2. Categorical
filtered_categorical_loans_df = filtered_loans_df.select_dtypes([object])
filtered_numeric_loans_df = filtered_loans_df.select_dtypes(['float64','int64'])

In [17]:
filtered_categorical_loans_df.head(5)

term int_rate grade sub_grade                 emp_title emp_length  \
0   36 months   10.65%     B        B2                       NaN  10+ years   
1   60 months   15.27%     C        C4                     Ryder   < 1 year   
2   36 months   15.96%     C        C5                       NaN  10+ years   
3   36 months   13.49%     C        C1       AIR RESOURCES BOARD  10+ years   
4   60 months   12.69%     B        B5  University Medical Group     1 year   

  home_ownership verification_status   issue_d  loan_status         purpose  \
0           RENT            Verified  Dec-2011   Fully Paid     credit_card   
1           RENT     Source Verified  Dec-2011  Charged Off             car   
2           RENT        Not Verified  Dec-2011   Fully Paid  small_business   
3           RENT     Source Verified  Dec-2011   Fully Paid           other   
4           RENT     Source Verified  Dec-2011   Fully Paid           other   

                  title zip_code addr_state earliest_cr_line revol_util  \
0              Computer    860xx         AZ         Jan-1985      83.7%   
1                  bike    309xx         GA         Apr-1999       9.4%   
2  real estate business    606xx         IL         Nov-2001      98.5%   
3              personel    917xx         CA         Feb-1996        21%   
4              Personal    972xx         OR         Jan-1996      53.9%   

  last_pymnt_d last_credit_pull_d debt_settlement_flag  
0     Jan-2015           Dec-2017                    N  
1     Apr-2013           Oct-2016                    N  
2     Jun-2014           Jun-2017                    N  
3     Jan-2015           Apr-2016                    N  
4     Jan-2017           Jan-2017                    N

In [18]:
# Print unique values for purpose and title
filtered_categorical_loans_df.purpose.value_counts()
filtered_categorical_loans_df.title.value_counts()

Debt consolidation                          648570
Credit card refinancing                     250516
Home improvement                             69997
Other                                        59637
Major purchase                               22082
Debt Consolidation                           15763
Medical expenses                             11882
Business                                     11280
Car financing                                10186
Moving and relocation                         7042
Vacation                                      6916
Consolidation                                 5385
debt consolidation                            4837
Home buying                                   4287
Debt Consolidation Loan                       3804
Credit Card Consolidation                     2360
consolidation                                 2175
Personal Loan                                 2133
Consolidation Loan                            1778
Home Improvement               

From the above you can see the title and purpose have similar values.
<p>'title' is the description of the purpose of the loan. So we can drop the tilte.</p>

<p>Similarly, zip_code and addr_state gives the same infomration. so we can drop zip_code</p>

<p>Also, subgrade is the sub category of grade.So we will drop subgrade column too


In [19]:
filtered_categorical_loans_df = filtered_categorical_loans_df.drop(['title','zip_code','sub_grade'],axis=1)

In [20]:
filtered_categorical_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 16 columns):
term                    1321847 non-null object
int_rate                1321847 non-null object
grade                   1321847 non-null object
emp_title               1241887 non-null object
emp_length              1321847 non-null object
home_ownership          1321847 non-null object
verification_status     1321847 non-null object
issue_d                 1321847 non-null object
loan_status             1321847 non-null object
purpose                 1321847 non-null object
addr_state              1321847 non-null object
earliest_cr_line        1321818 non-null object
revol_util              1321081 non-null object
last_pymnt_d            1320660 non-null object
last_credit_pull_d      1321784 non-null object
debt_settlement_flag    1321847 non-null object
dtypes: object(16)
memory usage: 171.4+ MB


In [21]:
#Now replace the missing values with mode
filtered_categorical_loans_df = filtered_categorical_loans_df.apply(lambda x:x.fillna(x.value_counts().index[0]))

In [22]:
#Now the number of missing values in each column should be zero
filtered_categorical_loans_df.isnull().sum().sort_values(ascending=False)

debt_settlement_flag    0
last_credit_pull_d      0
last_pymnt_d            0
revol_util              0
earliest_cr_line        0
addr_state              0
purpose                 0
loan_status             0
issue_d                 0
verification_status     0
home_ownership          0
emp_length              0
emp_title               0
grade                   0
int_rate                0
term                    0
dtype: int64

In [23]:
filtered_categorical_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 16 columns):
term                    1321847 non-null object
int_rate                1321847 non-null object
grade                   1321847 non-null object
emp_title               1321847 non-null object
emp_length              1321847 non-null object
home_ownership          1321847 non-null object
verification_status     1321847 non-null object
issue_d                 1321847 non-null object
loan_status             1321847 non-null object
purpose                 1321847 non-null object
addr_state              1321847 non-null object
earliest_cr_line        1321847 non-null object
revol_util              1321847 non-null object
last_pymnt_d            1321847 non-null object
last_credit_pull_d      1321847 non-null object
debt_settlement_flag    1321847 non-null object
dtypes: object(16)
memory usage: 171.4+ MB


In [24]:
filtered_categorical_loans_df.head()

term int_rate grade                 emp_title emp_length  \
0   36 months   10.65%     B                   Teacher  10+ years   
1   60 months   15.27%     C                     Ryder   < 1 year   
2   36 months   15.96%     C                   Teacher  10+ years   
3   36 months   13.49%     C       AIR RESOURCES BOARD  10+ years   
4   60 months   12.69%     B  University Medical Group     1 year   

  home_ownership verification_status   issue_d  loan_status         purpose  \
0           RENT            Verified  Dec-2011   Fully Paid     credit_card   
1           RENT     Source Verified  Dec-2011  Charged Off             car   
2           RENT        Not Verified  Dec-2011   Fully Paid  small_business   
3           RENT     Source Verified  Dec-2011   Fully Paid           other   
4           RENT     Source Verified  Dec-2011   Fully Paid           other   

  addr_state earliest_cr_line revol_util last_pymnt_d last_credit_pull_d  \
0         AZ         Jan-1985      83.7%     Jan-2015           Dec-2017   
1         GA         Apr-1999       9.4%     Apr-2013           Oct-2016   
2         IL         Nov-2001      98.5%     Jun-2014           Jun-2017   
3         CA         Feb-1996        21%     Jan-2015           Apr-2016   
4         OR         Jan-1996      53.9%     Jan-2017           Jan-2017   

  debt_settlement_flag  
0                    N  
1                    N  
2                    N  
3                    N  
4                    N

In [25]:
#The colums int_rate and revol_util are in percantage. we will convert it to numeric values.
filtered_categorical_loans_df['int_rate'] = filtered_categorical_loans_df['int_rate'].replace('%','',regex=True).astype('float')

filtered_categorical_loans_df['revol_util'] = filtered_categorical_loans_df['revol_util'].replace('%','',regex=True).astype('float')


In [26]:
filtered_categorical_loans_df.head()

term  int_rate grade                 emp_title emp_length  \
0   36 months     10.65     B                   Teacher  10+ years   
1   60 months     15.27     C                     Ryder   < 1 year   
2   36 months     15.96     C                   Teacher  10+ years   
3   36 months     13.49     C       AIR RESOURCES BOARD  10+ years   
4   60 months     12.69     B  University Medical Group     1 year   

  home_ownership verification_status   issue_d  loan_status         purpose  \
0           RENT            Verified  Dec-2011   Fully Paid     credit_card   
1           RENT     Source Verified  Dec-2011  Charged Off             car   
2           RENT        Not Verified  Dec-2011   Fully Paid  small_business   
3           RENT     Source Verified  Dec-2011   Fully Paid           other   
4           RENT     Source Verified  Dec-2011   Fully Paid           other   

  addr_state earliest_cr_line  revol_util last_pymnt_d last_credit_pull_d  \
0         AZ         Jan-1985        83.7     Jan-2015           Dec-2017   
1         GA         Apr-1999         9.4     Apr-2013           Oct-2016   
2         IL         Nov-2001        98.5     Jun-2014           Jun-2017   
3         CA         Feb-1996        21.0     Jan-2015           Apr-2016   
4         OR         Jan-1996        53.9     Jan-2017           Jan-2017   

  debt_settlement_flag  
0                    N  
1                    N  
2                    N  
3                    N  
4                    N

The columns issue_d ,earliest_cr_line,last_pymnt_d,last_credit_pull_d are in Mon-yyyy date format.We are going to create 2 numeric columns from each date by splitting into a month and year.

In [27]:
filtered_categorical_loans_df['issue_year'] = pd.DatetimeIndex(filtered_categorical_loans_df['issue_d']).year
filtered_categorical_loans_df['issue_month'] = pd.DatetimeIndex(filtered_categorical_loans_df['issue_d']).month

In [28]:
filtered_categorical_loans_df['earliest_cr_line_month'] = pd.DatetimeIndex(filtered_categorical_loans_df['earliest_cr_line']).month
filtered_categorical_loans_df['earliest_cr_line_year'] = pd.DatetimeIndex(filtered_categorical_loans_df['earliest_cr_line']).year


In [29]:
filtered_categorical_loans_df['last_pymnt_month'] = pd.DatetimeIndex(filtered_categorical_loans_df['last_pymnt_d']).month
filtered_categorical_loans_df['last_pymnt_year'] = pd.DatetimeIndex(filtered_categorical_loans_df['last_pymnt_d']).year

In [30]:
filtered_categorical_loans_df['last_credit_pull_month'] = pd.DatetimeIndex(filtered_categorical_loans_df['last_credit_pull_d']).month
filtered_categorical_loans_df['last_credit_pull_year'] = pd.DatetimeIndex(filtered_categorical_loans_df['last_credit_pull_d']).year


In [31]:
#lets drop issue_d ,earliest_cr_line,last_pymnt_d,last_credit_pull_d
filtered_categorical_loans_df = filtered_categorical_loans_df.drop(['issue_d','earliest_cr_line','last_pymnt_d','last_credit_pull_d'],axis=1)


In [32]:
filtered_categorical_loans_df['term'].value_counts()

 36 months    944664
 60 months    377183
Name: term, dtype: int64

In [33]:
#Convert the term column to numeruc by removing the months.
filtered_categorical_loans_df['term'] = filtered_categorical_loans_df['term'].replace('months','',regex=True).astype('int')


In [34]:
filtered_categorical_loans_df.emp_title.value_counts()

Teacher                                101680
Manager                                 18643
Owner                                   10383
Registered Nurse                         8906
RN                                       8638
Supervisor                               8200
Sales                                    7108
Driver                                   6829
Project Manager                          6346
Office Manager                           5530
General Manager                          5246
Director                                 5104
owner                                    4837
manager                                  4526
President                                4447
teacher                                  4276
Engineer                                 4262
Vice President                           3636
driver                                   3501
Operations Manager                       3396
Attorney                                 3274
Accountant                        

In [35]:
#There are total of 381900 unique values for emp_title. For now we can drop emp_title for our analysis
filtered_categorical_loans_df = filtered_categorical_loans_df.drop(['emp_title'],axis=1)

In [36]:
filtered_categorical_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 19 columns):
term                      1321847 non-null int64
int_rate                  1321847 non-null float64
grade                     1321847 non-null object
emp_length                1321847 non-null object
home_ownership            1321847 non-null object
verification_status       1321847 non-null object
loan_status               1321847 non-null object
purpose                   1321847 non-null object
addr_state                1321847 non-null object
revol_util                1321847 non-null float64
debt_settlement_flag      1321847 non-null object
issue_year                1321847 non-null int64
issue_month               1321847 non-null int64
earliest_cr_line_month    1321847 non-null int64
earliest_cr_line_year     1321847 non-null int64
last_pymnt_month          1321847 non-null int64
last_pymnt_year           1321847 non-null int64
last_credit_pull_month    1321847 non-null 

In [37]:
#Now create a 2 dataframes from filtered_categorical_loans_df
#1.With categorical values - Stage1 (This is used for data exploration)
#2.Converting categorical values to numerical values - Stage 2
filtered_categorical_loans_stage1_df = filtered_categorical_loans_df.copy(deep=True)
filtered_categorical_loans_stage2_df = filtered_categorical_loans_df.copy(deep=True)

In [38]:
filtered_categorical_loans_stage1_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 19 columns):
term                      1321847 non-null int64
int_rate                  1321847 non-null float64
grade                     1321847 non-null object
emp_length                1321847 non-null object
home_ownership            1321847 non-null object
verification_status       1321847 non-null object
loan_status               1321847 non-null object
purpose                   1321847 non-null object
addr_state                1321847 non-null object
revol_util                1321847 non-null float64
debt_settlement_flag      1321847 non-null object
issue_year                1321847 non-null int64
issue_month               1321847 non-null int64
earliest_cr_line_month    1321847 non-null int64
earliest_cr_line_year     1321847 non-null int64
last_pymnt_month          1321847 non-null int64
last_pymnt_year           1321847 non-null int64
last_credit_pull_month    1321847 non-null 

In [39]:
filtered_categorical_loans_stage2_df['loan_status'].value_counts()

Fully Paid                                             639250
Current                                                479981
Charged Off                                            170927
Late (31-120 days)                                      16750
In Grace Period                                          8304
Late (16-30 days)                                        3773
Does not meet the credit policy. Status:Fully Paid       1988
Does not meet the credit policy. Status:Charged Off       761
Default                                                   113
Name: loan_status, dtype: int64

In [40]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
filtered_categorical_loans_stage2_df['loan_status'] = le.fit_transform(filtered_categorical_loans_df['loan_status'])

In [41]:
filtered_categorical_loans_stage2_df['loan_status'].value_counts()

5    639250
1    479981
0    170927
8     16750
6      8304
7      3773
4      1988
3       761
2       113
Name: loan_status, dtype: int64

In [42]:
filtered_categorical_loans_stage2_df = pd.get_dummies(filtered_categorical_loans_stage2_df,drop_first=True)

In [43]:
filtered_categorical_loans_stage2_df.head()

term  int_rate  loan_status  revol_util  issue_year  issue_month  \
0    36     10.65            5        83.7        2011           12   
1    60     15.27            0         9.4        2011           12   
2    36     15.96            5        98.5        2011           12   
3    36     13.49            5        21.0        2011           12   
4    60     12.69            5        53.9        2011           12   

   earliest_cr_line_month  earliest_cr_line_year  last_pymnt_month  \
0                       1                   1985                 1   
1                       4                   1999                 4   
2                      11                   2001                 6   
3                       2                   1996                 1   
4                       1                   1996                 1   

   last_pymnt_year           ...            addr_state_TN  addr_state_TX  \
0             2015           ...                        0              0   
1             2013           ...                        0              0   
2             2014           ...                        0              0   
3             2015           ...                        0              0   
4             2017           ...                        0              0   

   addr_state_UT  addr_state_VA  addr_state_VT  addr_state_WA  addr_state_WI  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   addr_state_WV  addr_state_WY  debt_settlement_flag_Y  
0              0              0                       0  
1              0              0                       0  
2              0              0                       0  
3              0              0                       0  
4              0              0                       0  

[5 rows x 100 columns]

In [44]:
filtered_categorical_loans_stage1_df.head()

term  int_rate grade emp_length home_ownership verification_status  \
0    36     10.65     B  10+ years           RENT            Verified   
1    60     15.27     C   < 1 year           RENT     Source Verified   
2    36     15.96     C  10+ years           RENT        Not Verified   
3    36     13.49     C  10+ years           RENT     Source Verified   
4    60     12.69     B     1 year           RENT     Source Verified   

   loan_status         purpose addr_state  revol_util debt_settlement_flag  \
0   Fully Paid     credit_card         AZ        83.7                    N   
1  Charged Off             car         GA         9.4                    N   
2   Fully Paid  small_business         IL        98.5                    N   
3   Fully Paid           other         CA        21.0                    N   
4   Fully Paid           other         OR        53.9                    N   

   issue_year  issue_month  earliest_cr_line_month  earliest_cr_line_year  \
0        2011           12                       1                   1985   
1        2011           12                       4                   1999   
2        2011           12                      11                   2001   
3        2011           12                       2                   1996   
4        2011           12                       1                   1996   

   last_pymnt_month  last_pymnt_year  last_credit_pull_month  \
0                 1             2015                      12   
1                 4             2013                      10   
2                 6             2014                       6   
3                 1             2015                       4   
4                 1             2017                       1   

   last_credit_pull_year  
0                   2017  
1                   2016  
2                   2017  
3                   2016  
4                   2017

In [45]:
#Numeric columns
filtered_numeric_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 66 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
installment                   1321847 non-null float64
annual_inc                    1321843 non-null float64
dti                           1321782 non-null float64
delinq_2yrs                   1321818 non-null float64
fico_range_low                1321847 non-null float64
fico_range_high               1321847 non-null float64
inq_last_6mths                1321817 non-null float64
open_acc                      1321818 non-null float64
pub_rec                       1321818 non-null float64
revol_bal                     1321847 non-null float64
total_acc                     1321818 non-null float64
out_prncp                     1321847 non-null float64
out_prncp_inv                 1321847 non-null float64
total_py

In [46]:
filtered_numeric_loans_df.isnull().sum().sort_values(ascending=False)

mths_since_recent_inq         182177
num_tl_120dpd_2m              118999
mo_sin_old_il_acct            107535
pct_tl_nvr_dlq                 70429
avg_cur_bal                    70288
mo_sin_old_rev_tl_op           70277
mo_sin_rcnt_rev_tl_op          70277
num_rev_accts                  70277
num_bc_tl                      70276
tot_coll_amt                   70276
tot_cur_bal                    70276
total_rev_hi_lim               70276
mo_sin_rcnt_tl                 70276
num_accts_ever_120_pd          70276
num_actv_bc_tl                 70276
num_actv_rev_tl                70276
total_il_high_credit_limit     70276
num_il_tl                      70276
num_tl_30dpd                   70276
num_tl_op_past_12m             70276
num_tl_90g_dpd_24m             70276
tot_hi_cred_lim                70276
num_rev_tl_bal_gt_0            70276
num_op_rev_tl                  70276
bc_util                        63320
percent_bc_gt_75               63024
bc_open_to_buy                 62603
m

In [47]:
#Fill the missing values in each column with mean value
filtered_numeric_loans_df = filtered_numeric_loans_df.apply(lambda x: x.fillna(x.mean()),axis=1)

In [48]:
#The number of missing values in each columm should be zero
filtered_numeric_loans_df.isnull().sum().sort_values(ascending=False)

total_il_high_credit_limit    0
total_pymnt                   0
total_rev_hi_lim              0
tot_cur_bal                   0
tot_coll_amt                  0
acc_now_delinq                0
collections_12_mths_ex_med    0
last_fico_range_low           0
last_fico_range_high          0
last_pymnt_amnt               0
collection_recovery_fee       0
recoveries                    0
total_rec_late_fee            0
total_rec_int                 0
total_rec_prncp               0
total_pymnt_inv               0
out_prncp_inv                 0
total_bc_limit                0
out_prncp                     0
total_acc                     0
revol_bal                     0
pub_rec                       0
open_acc                      0
inq_last_6mths                0
fico_range_high               0
fico_range_low                0
delinq_2yrs                   0
dti                           0
annual_inc                    0
installment                   0
                             ..
total_ba

In [49]:
#There are four columns that has fico scores
#last_fico_range_low,last_fico_range_high,fico_range_low,fico_range_high
#Create a new column with average of all 4 columns.
filtered_numeric_loans_df['fico_average'] = (filtered_numeric_loans_df['fico_range_high'] + filtered_numeric_loans_df['fico_range_low']) / 2

In [50]:
#drop the four fico score columns
filtered_numeric_loans_df = filtered_numeric_loans_df.drop(['fico_range_high','fico_range_low','last_fico_range_high','last_fico_range_low'],axis=1)

In [51]:
filtered_numeric_loans_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 63 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
installment                   1321847 non-null float64
annual_inc                    1321847 non-null float64
dti                           1321847 non-null float64
delinq_2yrs                   1321847 non-null float64
inq_last_6mths                1321847 non-null float64
open_acc                      1321847 non-null float64
pub_rec                       1321847 non-null float64
revol_bal                     1321847 non-null float64
total_acc                     1321847 non-null float64
out_prncp                     1321847 non-null float64
out_prncp_inv                 1321847 non-null float64
total_pymnt                   1321847 non-null float64
total_pymnt_inv               1321847 non-null float64
total_re

In [52]:
final_filtered_stage1_df = pd.concat([filtered_numeric_loans_df, filtered_categorical_loans_stage1_df], axis=1)

final_filtered_stage1_df.to_csv('preprocessed_loans_stage1.csv')

In [53]:
final_filtered_stage1_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Data columns (total 82 columns):
loan_amnt                     1321847 non-null float64
funded_amnt                   1321847 non-null float64
funded_amnt_inv               1321847 non-null float64
installment                   1321847 non-null float64
annual_inc                    1321847 non-null float64
dti                           1321847 non-null float64
delinq_2yrs                   1321847 non-null float64
inq_last_6mths                1321847 non-null float64
open_acc                      1321847 non-null float64
pub_rec                       1321847 non-null float64
revol_bal                     1321847 non-null float64
total_acc                     1321847 non-null float64
out_prncp                     1321847 non-null float64
out_prncp_inv                 1321847 non-null float64
total_pymnt                   1321847 non-null float64
total_pymnt_inv               1321847 non-null float64
total_re

In [54]:
final_filtered_stage2_df = pd.concat([filtered_numeric_loans_df, filtered_categorical_loans_stage2_df], axis=1)

final_filtered_stage2_df.to_csv('preprocessed_loans_stage2.csv')

In [55]:
final_filtered_stage2_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1321847 entries, 0 to 103545
Columns: 163 entries, loan_amnt to debt_settlement_flag_Y
dtypes: float64(65), int64(10), uint8(88)
memory usage: 877.4 MB


<p> The total number of columns are 163 with 1321847 entries

<h4>Now lets explore the numeric values</h4>

In [1]:
filtered_numeric_loans_df.info()

NameError: name 'filtered_numeric_loans_df' is not defined